File: ui/src/app/actions.rs
use super::state::{AppState, NodeId, SpecRef};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum UiAction {
    SelectNode(NodeId),
    SelectNext,
    SelectPrevious,
    AddSiblingNode,
    IndentNode,
    OutdentNode,
    ToggleCollapse,
}

pub fn dispatch(state: &mut AppState, action: UiAction) -> bool {
    match action {
        UiAction::SelectNode(node_id) => {
            if state.selected_node == Some(node_id) {
                return false;
            }
            state.set_selected(node_id);
            true
        }
        UiAction::SelectNext => move_selection(state, 1),
        UiAction::SelectPrevious => move_selection(state, -1),
        UiAction::AddSiblingNode => add_sibling_node(state),
        UiAction::IndentNode => indent_selected(state),
        UiAction::OutdentNode => outdent_selected(state),
        UiAction::ToggleCollapse => state.toggle_collapse(),
    }
}

fn move_selection(state: &mut AppState, delta: isize) -> bool {
    let flattened = state.flattened_nodes();
    if flattened.is_empty() {
        return false;
    }

    let current = state
        .selected_node
        .and_then(|selected| flattened.iter().position(|row| row.id == selected))
        .unwrap_or(0);

    let target = if delta.is_negative() {
        current.saturating_sub(delta.unsigned_abs())
    } else {
        let next = current.saturating_add(delta as usize);
        next.min(flattened.len() - 1)
    };

    if current == target {
        return false;
    }

    state.set_selected(flattened[target].id);
    true
}

fn add_sibling_node(state: &mut AppState) -> bool {
    let parent = state
        .selected_node
        .and_then(|selected| state.nodes[selected].parent);

    // Generate a temporary spec_ref; will be replaced by backend on next sync.
    let tmp_ref: SpecRef = format!("local/new-{}", state.nodes.len());

    let new_id = state.create_node(tmp_ref, String::new(), parent);

    let insertion_index = if let Some(selected) = state.selected_node {
        state
            .siblings(parent)
            .iter()
            .position(|id| *id == selected)
            .map(|idx| idx + 1)
            .unwrap_or_else(|| state.siblings(parent).len())
    } else {
        state.siblings(parent).len()
    };

    state.siblings_mut(parent).insert(insertion_index, new_id);
    state.selected_node = Some(new_id);
    true
}

fn indent_selected(state: &mut AppState) -> bool {
    let Some(selected) = state.selected_node else {
        return false;
    };

    let parent = state.nodes[selected].parent;
    let sibling_snapshot = state.siblings(parent).to_vec();

    let Some(index) = sibling_snapshot.iter().position(|id| *id == selected) else {
        return false;
    };

    if index == 0 {
        return false;
    }

    let new_parent = sibling_snapshot[index - 1];

    state.siblings_mut(parent).remove(index);
    state.nodes[selected].parent = Some(new_parent);
    state.nodes[new_parent].children.push(selected);
    true
}

fn outdent_selected(state: &mut AppState) -> bool {
    let Some(selected) = state.selected_node else {
        return false;
    };

    let Some(parent) = state.nodes[selected].parent else {
        return false;
    };

    let grand_parent = state.nodes[parent].parent;

    let Some(index_in_parent) = state.nodes[parent]
        .children
        .iter()
        .position(|id| *id == selected)
    else {
        return false;
    };

    state.nodes[parent].children.remove(index_in_parent);

    let parent_index = state
        .siblings(grand_parent)
        .iter()
        .position(|id| *id == parent)
        .map(|idx| idx + 1)
        .unwrap_or_else(|| state.siblings(grand_parent).len());

    state
        .siblings_mut(grand_parent)
        .insert(parent_index, selected);
    state.nodes[selected].parent = grand_parent;
    true
}

File: archive/app/src/lib/stores/actions.ts
/**
 * UI action dispatcher.
 * Ported from ui/src/app/actions.rs.
 *
 * Pure state mutations that do NOT call the backend directly.
 * Backend RPCs are initiated by the connection manager based on returned signals.
 */

import type { NodeId, SpecRef } from '$types/index'
import { appState } from './app-state.svelte'

// ─── Action types ──────────────────────────────────────────────────────────────

export type UiAction =
  | { type: 'selectNode'; nodeId: NodeId }
  | { type: 'selectNext' }
  | { type: 'selectPrevious' }
  | { type: 'addSiblingNode' }
  | { type: 'indentNode' }
  | { type: 'outdentNode' }
  | { type: 'toggleCollapse' }
  | { type: 'enterEditing' }
  | { type: 'exitEditing' }

// ─── Dispatch ────────────────────────────────────────────────────────────────

/**
 * Dispatch a UI action against the global app state.
 * Returns true if the state actually changed (useful for dirty-checking).
 */
export function dispatch(action: UiAction): boolean {
  switch (action.type) {
    case 'selectNode':
      if (appState.selectedNode === action.nodeId) return false
      appState.setSelected(action.nodeId)
      appState.editorMode = 'selection'
      return true

    case 'selectNext':
      return moveSelection(1)

    case 'selectPrevious':
      return moveSelection(-1)

    case 'addSiblingNode':
      return addSiblingNode()

    case 'indentNode':
      return indentSelected()

    case 'outdentNode':
      return outdentSelected()

    case 'toggleCollapse':
      return appState.toggleCollapse()

    case 'enterEditing':
      if (appState.selectedNode === null) return false
      appState.editorMode = 'editing'
      return true

    case 'exitEditing':
      if (appState.editorMode !== 'editing') return false
      appState.editorMode = 'selection'
      return true
  }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function moveSelection(delta: number): boolean {
  // Skip code_ref sub-rows — arrow keys navigate only real spec nodes.
  const flattened = appState.flattenedNodes.filter((n) => n.kind !== 'codeRef')
  if (flattened.length === 0) return false

  const current = appState.selectedNode !== null
    ? flattened.findIndex((row) => row.id === appState.selectedNode)
    : 0

  const idx = current === -1 ? 0 : current
  const target = Math.max(0, Math.min(flattened.length - 1, idx + delta))

  if (idx === target && current !== -1) return false

  appState.setSelected(flattened[target].id)
  appState.editorMode = 'selection'
  return true
}

function addSiblingNode(): boolean {
  const selected = appState.selectedNode
  const parent = selected !== null ? (appState.nodes[selected]?.parent ?? null) : null

  // Generate a local temporary ref; will be replaced on next backend sync.
  const tmpRef: SpecRef = `local/new-${appState.nodes.length}`
  const newId = appState.createNode(tmpRef, '', parent)

  const siblings = appState.siblings(parent)
  const insertionIndex = selected !== null
    ? (siblings.indexOf(selected) + 1 || siblings.length)
    : siblings.length

  // Splice the new node into the sibling list.
  siblings.splice(insertionIndex, 0, newId)

  appState.setSelected(newId)
  appState.editorMode = 'editing'
  return true
}

function indentSelected(): boolean {
  const selected = appState.selectedNode
  if (selected === null) return false

  const parent = appState.nodes[selected]?.parent ?? null
  const sibs = appState.siblings(parent)
  const index = sibs.indexOf(selected)

  if (index <= 0) return false // no previous sibling to become the parent

  const newParent = sibs[index - 1]

  sibs.splice(index, 1)
  appState.nodes[selected].parent = newParent
  appState.nodes[newParent].children.push(selected)
  return true
}

function outdentSelected(): boolean {
  const selected = appState.selectedNode
  if (selected === null) return false

  const parent = appState.nodes[selected]?.parent ?? null
  if (parent === null) return false // already a root node

  const grandParent = appState.nodes[parent]?.parent ?? null
  const indexInParent = appState.nodes[parent].children.indexOf(selected)
  if (indexInParent === -1) return false

  appState.nodes[parent].children.splice(indexInParent, 1)

  const parentIndex = appState.siblings(grandParent).indexOf(parent)
  const insertAfter = parentIndex === -1
    ? appState.siblings(grandParent).length
    : parentIndex + 1

  appState.siblings(grandParent).splice(insertAfter, 0, selected)
  appState.nodes[selected].parent = grandParent
  return true
}
